src/Core/Repositories/SqlServer/OrganizationUserRepository.cs

```diff
@@ -281,5 +281,62 @@ public async Task DeleteManyAsync(IEnumerable<Guid> organizationUserIds)
                     new { Ids = organizationUserIds.ToGuidIdArrayTVP() }, commandType: CommandType.StoredProcedure);
             }
         }
+
+        public async Task UpsertManyAsync(IEnumerable<OrganizationUser> organizationUsers)
+        {
+            var createUsers = new List<OrganizationUser>();
+            var replaceUsers = new List<OrganizationUser>();
+            foreach (var organizationUser in organizationUsers)
+            {
+                if (organizationUser.Id.Equals(default))
+                {
+                    createUsers.Add(organizationUser);
+                }
+                else
+                {
+                    replaceUsers.Add(organizationUser);
+                }
+            }
+
+            await CreateManyAsync(createUsers);
+            await ReplaceManyAsync(replaceUsers);
+        }
+
+        public async Task CreateManyAsync(IEnumerable<OrganizationUser> organizationUsers)
+        {
+            if (!organizationUsers.Any())
+            {
+                return;
+            }
+
+            foreach(var organizationUser in organizationUsers)
+            {
+                organizationUser.SetNewId();
+            }
+
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var results = await connection.ExecuteAsync(
+                    $"[{Schema}].[{Table}_CreateMany]",
+                    new { OrganizationUsersInput = CoreHelpers.ToTVP("[dbo].[OrganizationUserType]", organizationUsers) },
+                    commandType: CommandType.StoredProcedure);
+            }
+        }
+
+        public async Task ReplaceManyAsync(IEnumerable<OrganizationUser> organizationUsers)
+        {
+            if (!organizationUsers.Any())
+            {
+                return;
+            }
+
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                var results = await connection.ExecuteAsync(
+                    $"[{Schema}].[{Table}_UpdateMany]",
+                    new { OrganizationUsersInput = CoreHelpers.ToTVP("[dbo].[OrganizationUserType]", organizationUsers) },
```

Comment:
In my experience (at least with insight.database, perhaps not Dapper?) you can just pass an `IEnumerable<T>` directly as the parameter and the auto-mapper will "do the right thing" with the UDT.

Reply:
I will try again -- It's possible I messed something up, but Unfortunately that wasn't working for me.

I need to read up on Dapper, I'm new to it. I definitely could have missed something

Reply:
Here's what I found https://dapper-tutorial.net/parameter-table-valued-parameter

It looks like conversion to a data table is exactly what's expected. That page doesn't specify it, but matching order is expected as well.

Reply:
okay, bummer, 🙁